---
title: "Exercise week 6"
author: "Alo√Øs THOMAS"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(GenomicRanges)
  library(ggplot2)
  #library(memes) # for the meme-based methods -- COMMENT OUT when using alternatives
  library(motifmatchr)
  library(Biostrings) # for handling sequences
  library(MotifDb) # database of motifs
  library(TFBSTools) # for handling some motif formats
  library(universalmotif) # for converting motifs to various formats
  library(PWMEnrich) # for R-based motif enrichment analysis
  library(Rsamtools)
  library(Biostrings)
  library(BSgenome)

})
ah <- AnnotationHub(localHub = FALSE)

# loading function findMotifInstances
#' findMotifInstances
#'
#' A wrapper around TFBSTools for scanning motif occurence, handling the
#' coordinate conversion as `memes` does.
#'
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with
#'   coordinate as names to enable conversion.
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i],
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```

# Loading transcription factor CREB1 peaks from Mus Musculus myocyte cell type

```{r, eval=FALSE}
download.file("https://www.encodeproject.org/files/ENCFF599FYT/@@download/ENCFF599FYT.bed.gz", "CREB1.bed.gz", mode = "wb")
CREB1_peaks<- rtracklayer::import("CREB1.bed.gz", format="narrowPeak")

seqlevelsStyle(CREB1_peaks) <- "Ensembl"
```

```{r}
peak_centers <- resize(CREB1_peaks, fix="center", width=100)
# we get the genome sequence:
ah <- AnnotationHub(localHub=FALSE)
genome <- ah[["AH49723"]]

# we get the sequences corresponding to the peak centers:
peak_seqs <- Biostrings::getSeq(genome, peak_centers)
peak_seqs
as.character(peak_seqs[1])
```

# Motif scanning

```{r}
motifs <- query(MotifDb, "CREB1")
names(motifs)
motif <- motifs[["Hsapiens-HOCOMOCOv10-CREB1_HUMAN.H10MO.A"]]
view_motifs(motif)
```

```{r}
#dir.create("GRCh38_genome")
#"export(import.2bit(genome), "GRCh38_genome/genome.fasta.gz", #compress=TRUE)
#bgzip("GRCh38_genome/genome.fasta.gz")
```

```{r}
#motif2 <- convert_motifs(motif, class="TFBSTools-PFMatrix")
#moi2 <- motifmatchr::matchMotifs(motif2, subject=peak_centers, genome=Rsamtools::FaFile("GRCh38_genome/genome.fasta.bgz"), out="positions")[[1]]

moi2 <- findMotifInstances(peak_seqs, motif)
```

```{r}
# peaks with motif:
peaks_w_motif <- CREB1_peaks[overlapsAny(CREB1_peaks,moi2)]
length(peaks_w_motif)
length(CREB1_peaks)
length(peaks_w_motif)/length(CREB1_peaks)
```

Of the length(CREB1_peaks) peaks, [length(peaks_w_motif)/length(CREB1_peaks)] contain a motif.

## Proportion of overlaps of motifs in genome and peak file

```{r}
Biostrings::writeXStringSet(rtracklayer::import(genome), "genome.fa")
```

```{r}
motifs_CREB1 <- findMotifInstances(genome, motif)
motif_w_peaks <- genome[overlapsAny(genome, motifs_CREB1)]
length(motifs_CREB1)
length(motif_w_peaks)/length(motifs_CREB1)
```

Of the length(motifs_CREB1) motif instances, [length(motif_w_peaks)/length(motifs_CREB1)] overlap a peak.
